### PR TITLE
Modify defineVote

### PR DESCRIPTION
**Pull request by Skyvell**
Modified defineVote function to be callable by user.
Conditions:
- 20% quorum.
- 100 bnUSD fee.
- 0.1% of total baln supply staked.

Added gov the following gov parameters along with setters and getters:
- minimum_vote_duration
- baln_vote_definition_criteria
- bnUSD_vote_definition_criteria
- quorum

Additional changes not originally apart of the PR:
- Refund fee when vote canceled or passed.
- evaluateVote (extended executeVoteActions).
- Updated access of cancelVote and activateVote to owner or proposer.

Added unit tests and updated existing tests to be compatible with the new function.

### CODE REVIEW

File: core_contracts/governance/governance.py
Comment (simusud) [inline core_contracts/governance/governance.py]: Its better not to use float in the score completely as float calculation will get rejected in the score.
For the percentage we have been implementing it in terms of POINTS(= 10000) in loan score maybe you have take the reference from loans score.

**Review by simusud — CHANGES_REQUESTED** [review]

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Got it. I'll take a look at the loan contract and make those changes. Just to make sure I understand you correctly: 100% would be 10000 points and 0.01% would be 1 point, correct?

Comment (Skyvell) [reply core_contracts/governance/governance.py]: I'm not entirely sure what you mean by rejected? Audit? I talked to Spl3en (one of the contract auditors) regarding contracts failing contract audit because of float operations (I found it a bit wierd since float operations are entirely deterministic, so it should not be a problem). He confimed that contracts do not fail audit because of float operations, except when it's used as a return value for @external(readonly=true) functions or as a key dict type.

With that said. I'll of course change the code if you want it represented only with integers.

Comment (simusud) [reply core_contracts/governance/governance.py]: Yeah, since we were implementing percentage in terms of points all over the system i wanted it to be that way. And thanks for those insights that might  be useful in future.

**Review by simusud — DISMISSED** [review]
Looks good

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: That makes perfect sense. Thanks for reviewing 👍

File: core_contracts/governance/governance.py
Comment (Skyvell) [inline core_contracts/governance/governance.py]: After reviewing this one last time I realized that this call to the bnusd contract will make the governance contract the sender, and not the user. 

I see two different solutions here:

1. Use tokenfallback like you do in the dex contract when a user sends irc2 tokens to it. Pass method and arguments via the data field in the bnUSD contract. Then call defineVote in the tokenFallback method.
2. Add an additional transfer method in the bnUSD contract accessible only via the governance contract.

Something like: 

	@external
	@only_governance
	def govTransfer(self, _from: Address, _to: Address, _value: int, _data: bytes = None):
		if _data is None:
			_data = b'None'
		self._transfer(_from, _to, _value, _data)

Then pass self.msg.sender to this method when called from the governance contract.
Option 2 looks cleaner and easier to implement. Thoughts?

Comment (simusud) [reply core_contracts/governance/governance.py]: Yeah option 2 looks good and easier. we can implement that.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: It is a small thing, but criteria is the plural of criterion, and this is just one criterion. The code would read better if the variable name follows proper grammar.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: Also just a small thing: Please use the constant 'EXA' for the 10**18 value.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: We have been trying to follow the docstring convention used in most of the ICON code, an example of which is here - https://github.com/icon-project/icon-service/blob/master/iconservice/iconscore/icon_score_base.py#L649

Single line descriptions can also go on the same line as the opening triple quote, but that is a convention that they seem to go by about 50% of the time. Not sure offhand if it is preferred in Python style in general. Python style guide give a spec for multiline docstrings here - https://www.python.org/dev/peps/pep-0257/#id17 I'd refer to that in general, but also simply opt for consistency throughout.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: Thanks for improving the docstring here.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: vote_start is the day on which voting is first allowed so it also can't be equal to the current day, which started at the last snapshot time, so the comparison here needs to be <=.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py:176]: I was not sure if we necessarily had to restrict the snapshot to be the current day or later. A snapshot from a previous day could be valid, and it might make sense to allow the most recently completed snapshot, although I don't see any cases where an earlier one would be justified. In any case there has to be some limit. That most recently completed snapshot may actually be the best choice. Makes me wonder if we ought to provide defaults for snapshot, duration and vote_start.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [reply core_contracts/governance/governance.py]: Good catch. Yeah, I'd agree that is the better design for this case. A govTransfer method may be used for some of the other governance functions we will be adding as well. Please add that to the PR.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: Let's please remove this line. No need to set this again.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [reply core_contracts/governance/governance.py]: Let's also put these lines in a method scoreUpdate_12() so that the actions taken on update are documented.

**Review by dbrehmer — CHANGES_REQUESTED** [review]
Looks good. Just a few changes suggested.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Got it. Editing.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Got it. Wasn't sure if it still had some purpose. Removing.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Will do.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Sounds good. I'll make changes to fit that convention.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: No problem :)

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: I'll do that. Thanks.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Not sure I understand the reason here. You do not want users to be able to define a vote which starts at the same day as it was defined on? thus the <=? Is that understanding correct?

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py:176]: I agree that an earlier snapshot than current day might make sense in some rare cases. Perhaps in a senario where only users who used the platform at an earlier date are affected by some vote. Can't think of anything concrete right now though. And as you said, there has to be some limit.

Yeah, good question. If we go with defaults; would it make more sense to handle that via the frontend or backend? I'm not sure how you envision the vote creation process on the frontend, so it's hard for me to say.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [reply core_contracts/governance/governance.py]: Yeah, so the day change is at 17:00 UTC, if someone defines a vote at 23:00 UTC and we accept a vote_start equal to the current day, then 6 hours of voting time would already have passed. It is a little confusing and I am not sure how we could make it clearer. I think the difficulty comes from the snapshot time being at the end of a day, but if you say the vote starts on a given day it makes sense that it starts at the beginning of the day, which is the time of the snapshot for the previous day. I guess we could say the vote starts at a particular snapshot time, so if a vote is defined to start at snapshot 102 voting would actually be allowed when day 103 starts. As it is when you look at the data for a vote definition it shows an end day and votes are not accepted on that day, sort of like the range function in Python.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [reply core_contracts/governance/governance.py:176]: I guess I'd be good with just allowing the most recent completed snapshot or later for the reference.

For defaults I have tried to supply sensible defaults on contracts, but I guess it only really helps to make the frontend code a little cleaner, or maybe helps us if we are calling the methods directly. In general defaults could all be handled in the front end. That's probably good enough. So, yeah, we can skip adding defaults.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Ah, I see. Thank you for the detailed explanation.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py:176]: Alright. Sounds good.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Pushed some new changes to accommodate Dans review. Github says "Skyvell dismissed simusud’s stale review via 0fff5f1 3 hours ago", not sure why it said that; I did not dismiss anything :)

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Noticed a fix was deployed regarding the vote actions not being submitted in the right format. Changing this to scoreUpdate_13().

File: core_contracts/governance/governance.py
Comment (Skyvell) [inline core_contracts/governance/governance.py]: Coupled refunding to a successful execution of vote actions here. Not sure if you have something else in mind? Do we want a user to be refunded even if he makes a mistake when defining the actions?  One option could be to remove the revert statement here if we want the user to be refunded on a successful vote even if execution fails.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: I think we need to call this the percentage of the total eligible baln which must participate, rather than the total baln supply.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: Hmm, actually the snapshot has to be less than the vote_start because the snapshot for the vote_start day does not happen until the end of the day. Seems like it may not actually matter that much, but people could change their stake up to 24 hours after voting starts. If there were defaults the snapshot would be one less than the vote_start.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [reply core_contracts/governance/governance.py]: Yeah, seems like the refund should happen on a successful vote even if execution fails. If it is successful we will have to make it happen one way or another.

Comment (dbrehmer) [reply core_contracts/governance/governance.py]: One other thought about the process - We have this idea of activating a vote, or maybe we need to call it something else, and canceling a vote if there is something wrong with it. Do we want to allow that as part of the process? It should not include requiring the governance owner to approve. The activate or cancel would be called by the one who defined the vote. It would just give them the chance to back out if there was an error. If we aren't going to use those methods they could be removed. If they will be used we could make the payment happen in the activate, or approve, method. Would it even make sense to require an approver who is also qualified to submit a vote? Like a code review?

**Review by dbrehmer — CHANGES_REQUESTED** [review]
Looking good. Just a couple more things to consider. Only one small actual code change requested.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Yeah, agreed. Will fix.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Alright, editing.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: I see Scott just made a comment to remove the activateVote methods; so I guess I'll do that. I'll also add refunding to the cancelVote function along with access to the proposer.

Actually I see some problems with the executeVoteAction method. A proposals vote actions can be called repeatedly to execute already executed vote actions (as far as I can tell). Someone might call it on an old proposal that has already been executed and change some really vital stuff. I'll address this.

Another thing. You'll have two different kind of votes - votes that require execution and votes that are only "polls", like the one currently ongoing. But the proposal status "successful" is never asigned in any senario. I'll fix that too.

Not sure if an approval is nessasary, I guess the "pending" period could serve as a chance to get the review done, like you said?

Please let me know if I misunderstood something.

Edit: Removing activateVote creates some other issues, requiring even more changes. Holding off there.

Comment (Skyvell) [reply core_contracts/governance/governance.py]: Had to make a bunch of changes to fit it all together .. but should be pretty good now I think.

Comment (Skyvell) [reply core_contracts/governance/governance.py]: Optionally we could remove activateVote and activate the vote on creation, but allow it to be cancelled as long as it has not started yet. I was also thinking about "malicious" proposals. Like if you submit a proposal with clearly malicious parameter changes, payout of funds from treasury and such, but with a completely different description to trick users into voting yes. Very hard to spot for the average user. Would it be good to perhaps limit vote_start to >= current_day + 3 (or something like that). So there is ample time for someone to catch those and inform?

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py:151]: Now that cancel can be done from the active status we will also need to check that the vote has not started. I don't think it is right to allow a proposer to cancel a vote after people have started voting. If you agree, let's add a check whether the vote has started.

File: core_contracts/governance/governance.py
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: In the cases where we have the vote name as a parameter let's change it to the vote index. I had originally thought that using the name would make it more human readable, but that is not necessary and now we have some using the index and some using the name. My mistake to have used the name in the first place.

**Review by dbrehmer — CHANGES_REQUESTED** [review]
Looking good. Just two minor items to change.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py:151]: Yeah I agree. Adding that restriction for the proposer.

File: core_contracts/governance/governance.py
Comment (Skyvell) [reply core_contracts/governance/governance.py]: Yeah, I was not sure which one you preferred. Like you, I also thought it would be more readable with the name. Making the changes.

**Comment by Skyvell** [comment]
Thanks for the review. I made all changes but left castVote with parameter name. It has vote_index in development branch already.

**Review by dbrehmer — DISMISSED** [review]
Looks good. Thanks!

**Review by dbrehmer — APPROVED** [review]

**Pull request merged**
